### PR TITLE
Implementation of gitlab-oauth2 service for fossology

### PR DESCRIPTION
**Pull request by shaheemazmalmmd**
closses #515

### CODE REVIEW

**Comment by maximilianhuber** [comment]
With this changes most tests fail with `Segmentation fault`, see screenshot:

tmp

Additionally, all travis jobs were restarted, since at least one test failed due to a race condition.

**Comment by shaheemazmalmmd** [comment]
Error_(PHP Parse error:  syntax error, unexpected '[' in /home/travis/build/fossology/fossology/src/vendor/guzzlehttp/psr7/src/functions.php on line 78)_ seems to be with the package guzzlehttp. which is getting downloaded internally by composer. there is some issue created here as well https://github.com/guzzle/guzzle/issues/644
and 
Only PHP 5.5 (>=5.5.9) or greater version will support Oauth2. May be we should remove tests in travis which are executed for php less than 5.5.9 ??.

**Comment by maximilianhuber** [comment]
I think, we first have to finish the discussion in #585 before we drop testing of some php versions.

File: src/www/ui/core-auth.php [hunk 1]
```diff
@@ -214,8 +242,13 @@ public function Output()
         $referrer = Traceback_uri();
       }
     }
+    if(!empty($getEmail) && empty($userName)){
+      $oAuthClient = true; // To test in the server
```
Comment (maximilianhuber) [inline src/www/ui/core-auth.php]: This might be a line, which should be removed.

File: src/www/ui/core-auth.php [hunk 2]
```diff
@@ -266,15 +305,15 @@ function OutputOpen()
    *
    * @return boolean
    */
-  function checkUsernameAndPassword($userName, $password)
+  function checkUsernameAndPassword($userName, $password, $oAuthClient=false)
```
Comment (maximilianhuber) [inline src/www/ui/core-auth.php]: A function called `checkUsernameAndPassword` should not modify the global state e.g. `$_SESSION`

File: src/www/ui/core-auth.php [hunk 3]
```diff
@@ -202,6 +208,28 @@ public function Output()
     $userName = GetParm("username", PARM_TEXT);
     $password = GetParm("password", PARM_TEXT);
     $referrer = GetParm("HTTP_REFERER", PARM_TEXT);
+    $oAuthClient = GetParm("oAuthClient", PARM_TEXT);
+    $getEmail = GetParm("HTTP_SCMAIL", PARM_TEXT);
+    $providerCheck = GetParm("providerCheck", PARM_TEXT);
+
+    if(!empty($providerCheck) && empty($getEmail)){
```
Comment (maximilianhuber) [inline src/www/ui/core-auth.php]: This could be moved to another function

File: src/www/ui/core-auth.php [hunk 4]
```diff
@@ -202,6 +208,28 @@ public function Output()
...
+      $provider = new Gitlab([
+        "clientId"                => $SysConf['SYSCONFIG']['GitlabAppIdOauth'],
+        "clientSecret"            => $SysConf['SYSCONFIG']['GitlabSecretOauth'],
+        "redirectUri"             => $SysConf['SYSCONFIG']['RedirectOauthURL'],
+        "domain"                  => $domainOauth
+      ]);
+      $authorizationUrl = $provider->getAuthorizationUrl();
+      $_SESSION['oauth2state'] = $provider->getState();
+      header('Location: ' . $authorizationUrl);
+      exit;
```
Comment (maximilianhuber) [inline src/www/ui/core-auth.php]: Why do you exit here in such a harsh way?

**Comment by bufferoverflow** [comment]
@shaheemazmalmmd please fix the issues mentioned by @maximilianhuber and rebase so this can get merged

**Comment by mcjaeger** [comment]
we will do a relaunch here with a new PR

**Pull request closed without merging**